### ISSUE TITLE
show() doesn't open an interactive Matplotlib window

### ISSUE THREAD

**Issue by dstansby** [yt-project/yt#2126]
### Bug report

**Bug summary**

Calling show on a slice object isn't showing an interactive mpl window. Just to double check it isn't a backend issue, I've tried to explicitly set an interactive backend in the snippet below. 

```python
import matplotlib
import numpy as np
import yt

matplotlib.use('qt5agg')
yt.toggle_interactivity()


@yt.derived_field(name = "funfield", units="cm", take_log=False)
def funfield(field, data):
    return (np.sin(data["phi"])**2 + np.cos(data["theta"])**2) * (1.0*data["r"].uq+data["r"])


ds = yt.load_uniform_grid({}, [128, 128, 128],
                          bbox=np.array([[0.0, 1.0], [0.0, np.pi], [0.0, 2*np.pi]]),
                          geometry="spherical")


s = ds.slice(2, np.pi/2)
p = s.to_pw("funfield", origin="native")
p.set_zlim("all", 0.0, 4.0)
p.show()
```

**Actual outcome**
No window is shown, and the program exits normally.

**Expected outcome**
An interactive Matplotlib window is shown.

**Version Information**

  * Operating System: MacOS
  * Python Version: 3.7.1
  * yt version: 4.5.0 (conda-forge)
  * Other Libraries (if applicable): Matplotlib 3.0.2

**Comment by ngoldbaum** [comment]
So it *does* work inside of an IPython session using the `%run` magic. I can confirm it doesn't work if you run the script non-interactively.

It also works if I paste the code directly into a python interpreter.

What I think is happening is that an interactive plot window *is* getting created, but the python interpreter exits immediately afterwards. You can force the interpreter to stay open until you hit a key by e.g. adding a call to `input()` at the end of your script.

That said, the following script does block the interpreter from exiting, so presumably we should be able to do the same thing:

```
import matplotlib
matplotlib.use('qt5agg')
from matplotlib import pyplot as plt

plt.plot([1, 2, 3], [1, 2, 3])

plt.show()
```

Hope the workarounds I suggested above are sufficient for now and thank you for the report!

**Comment by dstansby** [comment]
I wonder if it's because there's calls to `fig.show()` inside yt? I remember having problems with that vs. `plt.show()` in the past (they should in theory do the same thing, but it might be a Maptlotlib bug). I'll have a dig and see what I can find.

**Comment by dstansby** [comment]
(p.s. thanks for the suggested workarounds!)

**Comment by ngoldbaum** [comment]
Yes, it's almost certainly something like that. yt hooks into matplotlib at a lower level than pyplot, so we need to do some extra work sometimes to emulate the behavior of pyplot.

In principle we don't actually need to hook into matplotlib at a low level anymore because matplotlib 3.0.0 recently grew the ability to automatically detect an appropriate backend at runtime and will no longer crash if you try to make a plot on a headless session out of the box (the original main justification for not using pyplot in yt as much as possible). That said, refactoring yt to make use of pyplot instead of the lower level figure and canvas objects would be a substantial undertaking.
